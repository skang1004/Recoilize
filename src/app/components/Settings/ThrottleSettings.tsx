import React, {useState} from 'react';

interface ThrottlesettingsProps {
  throttleDisplay: string;
  setThrottleDisplay: React.Dispatch<React.SetStateAction<string>>;
}

const ThrottleSettings: React.FC<ThrottlesettingsProps> = ({
  throttleDisplay,
  setThrottleDisplay,
}) => {
  const [throttleNum, setThrottleNum] = useState<string>('');

  // onClick function for reset button. 70ms is the default throttle
  const onClick = (): void => {
    setThrottleDisplay('70');
    setThrottleNum('70');
  };

  // Creating function to tie to get to the backend
  const throttleFunc = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    const backgroundConnection = chrome.runtime.connect();
    // post the message with index in payload to the connection
    backgroundConnection.postMessage({
      action: 'throttleEdit',
      tabId: chrome.devtools.inspectedWindow.tabId,
      payload: {value: throttleNum}, // edit this value to some other number
    });

    setThrottleDisplay(throttleNum);
    setThrottleNum('');
  };

  return (
    <div>
      <h2>Enter Throttle</h2>
      <form onSubmit={throttleFunc}>
        <input
          type="text"
          style={{marginBottom: '10px'}}
          onChange={e => setThrottleNum(e.target.value)}
          value={throttleNum}
          placeholder="enter in milliseconds"
        />{' '}
        <span style={{fontSize: '14px'}}>milliseconds</span>
        <div>
          <button type="submit">Enter</button>
          <button onClick={onClick} type="submit">
            Reset
          </button>
        </div>
      </form>
      <span>
        Current throttle is{' '}
        {throttleDisplay === '70'
          ? `${throttleDisplay} (default)`
          : throttleDisplay}{' '}
        ms
      </span>
    </div>
  );
};

export default ThrottleSettings;
