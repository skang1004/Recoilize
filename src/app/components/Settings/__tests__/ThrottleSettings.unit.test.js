import React, {useState} from 'react';
import ReactDOM from 'react-dom';
import {getQueriesForElement, getByText} from '@testing-library/dom';

import {render, fireEvent} from '@testing-library/react';

import ThrottleSettings from '../ThrottleSettings';

// Check the render
describe('Throttle Component renders correctly', () => {
  beforeEach(() => {
    const {} = render(<ThrottleSettings throttleDisplay={1000} />);
  });
  it('renders without crashing', () => {
    const root = document.createElement('div');
    ReactDOM.render(<ThrottleSettings />, root);
  });
  it('renders to the dom with the correct content', () => {
    const root = document.createElement('div');
    ReactDOM.render(<ThrottleSettings throttleDisplay={1000} />, root);
    const {getByText} = getQueriesForElement(root);
    expect(getByText('Enter Throttle')).not.toBeNull();
    expect(getByText('Enter')).not.toBeNull();
    expect(getByText('Reset')).not.toBeNull();
    expect(getByText('Current throttle is 1000 ms')).not.toBeNull();
  });
});

describe('Throttle Snapshots Testing', () => {
  it('renders & matches snapshots', () => {
    const {asFragment} = render(<ThrottleSettings />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Check button and user input functionalities', () => {
  it('Check the user input typing', () => {
    const {getByPlaceholderText} = render(
      <ThrottleSettings throttleDisplay={1000} />,
    );
    const input = getByPlaceholderText('enter in milliseconds');

    fireEvent.change(input, {target: {value: 500}}); // this successfully changes the value
    expect(input.value).toEqual('500');
  });

  it('Check the enter and reset button', () => {
    // creating mock functions for chrome
    chrome.runtime.connect = function () {
      return {postMessage: function () {}};
    };
    let setThrottleDisplay = jest.fn();
    chrome.devtools = {inspectedWindow: {}};

    // Rendering the component
    const {getByText} = render(
      <ThrottleSettings
        throttleDisplay={1000}
        setThrottleDisplay={setThrottleDisplay}
      />,
    );
    // Testing the buttons
    const submitButton = getByText('Enter'); // check these buttons -- how to test
    const resetButton = getByText('Reset');
    fireEvent.click(submitButton);
    fireEvent.click(resetButton);
  });
});
