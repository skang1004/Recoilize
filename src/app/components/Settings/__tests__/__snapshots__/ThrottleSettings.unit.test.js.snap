// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Throttle Snapshots Testing renders & matches snapshots 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Enter Throttle
    </h2>
    <form>
      <input
        placeholder="enter in milliseconds"
        style="margin-bottom: 10px;"
        type="text"
        value=""
      />
       
      <span
        style="font-size: 14px;"
      >
        milliseconds
      </span>
      <div>
        <button
          type="submit"
        >
          Enter
        </button>
        <button
          type="submit"
        >
          Reset
        </button>
      </div>
    </form>
    <span>
      Current throttle is  ms
    </span>
  </div>
</DocumentFragment>
`;
